File: checks/python/src/section508/checks/heading_order.py
from section508.dom import inclusive_select, tag_name

HEADING_SELECTOR = "h1, h2, h3, h4, h5, h6"


def check(root):
    headings = inclusive_select(root, HEADING_SELECTOR)
    details = []
    if not headings:
        return details

    previous = heading_level(headings[0])
    for heading in headings[1:]:
        level = heading_level(heading)
        if level > previous + 1:
            details.append(f"This section goes from h{previous} to h{level}.")
        previous = level
    return details


def heading_level(node):
    name = tag_name(node)
    try:
        return int(name[1:])
    except (TypeError, ValueError):
        return 0

File: checks/ruby/lib/section508/checks/heading_order.rb
# frozen_string_literal: true

module Section508
  module Checks
    # Do not skip a heading level inside the given node.
    class HeadingOrder
      SELECTOR = "h1, h2, h3, h4, h5, h6"

      def self.call(root)
        headings = headings_in(root)
        return [] if headings.empty?

        details = []
        previous = level(headings.first)
        headings.drop(1).each do |node|
          current = level(node)
          details << "This section goes from h#{previous} to h#{current}." if current > previous + 1
          previous = current
        end
        details
      end

      def self.headings_in(root)
        return [] unless root

        list = []
        list << root if Dom.element?(root) && Dom.matches?(root, SELECTOR)
        list.concat(Dom.elements(root, SELECTOR))
        list
      end

      def self.level(node)
        node.name.to_s[1].to_i
      end
    end
  end
end
